### DISCUSSION TITLE
Question regarding level crossings and other upcoming layout features.

### DISCUSSION THREAD

**Discussion by TekyTeky** [Novakasa/brickrail#149] [General]
Hi, about a week ago the monitor I had at the back of the layout for Brickrail came tumbling down and destroying part of the raised area and track etc... lucky it is Lego !! Hnece not much train running.
But as a result I thought I need to update the track to accomodate Brickrail more... and when I saw your Action regarding rail crossings I thought I need to plan for them... do you have a track length for the design of the crossings or block lengths etc?

Also I have ordered some track from TrixBrix, which are due this week. I am reviewing your BlueBrick layout, as I have purchased the Stadium Loop, some switches and curved switches. And wondered if you can send me your Layout.brl just so I can get an idea of where you have put your sensors, especially for the track going between layers? Cheers.

**Comment by Novakasa** [comment]
Aw, that's unfortunate about your layout. New beginnings, then!

The crossing motors are now implemented, see https://github.com/Novakasa/brickrail/issues/146 for details, I also posted a picture of my design for a crossing that only uses one motor. But this is just my design, for brickrail it doesn't really matter what your crossing looks like, you can also assign 2 motors to a crossing. In the virtual layout, it just is a single track that is assigned to a crossing, and how the crossing is built is up to you. It doesn't require additional markers, it just uses the fact that tracks have to "lock" a track section before they go that route, so every time the crossing track is "locked" the crossing booms will close.

For blocks, the main constraint is still that every train should fit between the two markers of every block. If you need prior markers, make sure that your layout has good spots for you to put them. The distance between the prior and the first block marker should ideally be roughly the same as the distance between the two block markers. But if your trains will go slowly, the distance can be shorter.

In the future I do plan to make it possible to have multiple priors per logical block. This should help in the case of a switch right before a block, as there are multiple ways the train can arrive from.

The most recent version of the layout file of my large layout is actually part of this repository: https://github.com/Novakasa/brickrail/blob/master/brickrail-gui/brickrail-layouts/ro_schattenbahnhof_umkehren_alt3_colors.brl
To download, click this button in the top right:
![image](https://github.com/Novakasa/brickrail/assets/59559612/c01a41a6-efdf-412b-b653-29bb37c236b8)

If you plan to have multiple layers, the ramp can be quite tricky to implement. I use one block per ramp, and I disallow trains to stop or reverse at these blocks in the block settings. For the markers, I set the "marker speed" to fast when the train starts to go uphill, and set the speed to slow when it starts to go downhill. This is required for the train to not derail after downhill or not stall when going uphill. You also need the ramp to be quite large, my layout is probably the minimum possible length for this height, as my heaviest train doesn't really make it with 3 wagons anymore. You also need better grip on the motor wheels, I use the rubber rings from trixbrix.

**Reply by TekyTeky** [reply]
Right, it all looks great and I’m excited to get into it. I was going to ask if you used the rubber rings.
It seems like you have been doing a lot, I also saw something about Godot4?
I hope you are on your summer break and not skipping classes 🫣

**Reply by Novakasa** [reply]
I'm not doing nearly as much as a few months ago, just some evenings when I have time.

Godot 4 was just an experiment, I ran the automatic conversion tool and there is a lot that I would have to fix for it to work, so seems unlikely at the moment.
